### ISSUE TITLE
Conditional execution

### ISSUE THREAD

**Issue by s1na** [Meta-tx/Harbour-MVP#11]
Hey everyone,

Meta txes are supposed to be submitted to the network immediately by relayers. It'd be however possible to delay their execution, and make it conditional, e.g. on [time](https://github.com/ethereum-alarm-clock/ethereum-alarm-clock), [state](https://blog.chronologic.network/time-is-only-the-beginning-introducing-conditional-scheduling-with-chronos-protocol-76036e4daaad?source=linkShare-3afeb2943a89-1538076828) or [events](https://github.com/planet-ethereum/ethbase). This could allow "automating" actions like revealing votes and bids, or claiming rewards, etc.

There are two main questions however:
- Whether they can be added to the meta tx standard, with an acceptable complexity overhead. It'd be amazing to have insight from @kosecki123 or @lsaether on this.
- Whether it's desirable to add them to this standard, or they should rather be a different standard.

**Comment by s1na** [comment]
**TL;DR** We discuss adding `(uint conditionType, bytes condition)` to executable signed messages, and employing these fields to enable meta txes conditioned on EVM events.

*Note:* Conditional execution is costlier in terms of gas, but this shouldn't add much overhead to non-conditional txes.

### Modifications to executable signed messages
As there are multiple ways to make transactions conditional, including by time, by checking state or by occurence of events, in addition to the data necessary for specifying the condition, we could add a condition type, which specifies how to parse the condition data.

#### Condition type
ConditionType | Purpose
--- | ---
`0x0` | Not conditional (default)
`0x1` | Time-based condition
`0x2` | EventLog occurence

#### Condition data
For EventLogs, the condition could be the RLP-encoded concatenation of `(address emitter, bytes32 eventTopic)`.

### Relayer
Relayers would store conditional meta txes, and when the event specified by one of the meta txes is emitted, they [compute a proof](https://github.com/planet-ethereum/relay-network/blob/master/ethbase/proof.go) for the receipt that contains the event log, and send it along with the block header and index of log in receipt to the smart contract.

### On-chain verification
For an incoming meta transaction which has a `conditionType == 0x2`:

- Block header sent by relayer is parsed to get block number and `receiptHash`.
- Hash of block header should be equal to `blockhash(blockNumber)`.
- Proof sent by relayer for receipt should be verified against `receiptHash`.
- Log is extracted from receipt (according to `logIndex`).
- `emitter` and `eventTopic` should match those of extracted log.
- Smart contract executes meta transactions.

Example for [verifying proof](https://github.com/planet-ethereum/ethbase/blob/master/contracts/Ethbase.sol#L72).
